### PR TITLE
[FIX] web: fieldDependency should be in fields too

### PR DESCRIPTION
**Pull request by kebeclibre**
Commit f45edfbe6d03b15977392dda9561cd7a3e16ce34 fixed a bug on image fields
but was a partial fix.

The problem is that while adding field dependencies to fieldsInfo,
it did not add them in the fields of the view, making the view crash.

After this commit, views and subviews contain fieldDependency as both info and actual fields of the view

OPW 780021, 781620, 779649

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by kebeclibre** [comment]
Much cleaner than #20793 , hence closing that one

File: addons/web/static/src/js/services/data_manager.js [hunk 1]
```diff
@@ -406,8 +406,12 @@ return core.Class.extend({
                 if (fieldsInfo[node.attrs.name].fieldDependencies) {
                     var deps = fieldsInfo[node.attrs.name].fieldDependencies;
                     for (var dependency_name in deps) {
+                        var dependency_dict = {'name': dependency_name, 'type': deps[dependency_name].type, 'options':  deps[dependency_name].options || {}};
                         if (!(dependency_name in fieldsInfo)) {
-                            fieldsInfo[dependency_name] = {'name': dependency_name, 'type': deps[dependency_name].type, 'options':  deps[dependency_name].options || {}};
+                            fieldsInfo[dependency_name] = dependency_dict;
+                        }
+                        if (!(dependency_name in fields)) {
+                            fields[dependency_name] = dependency_dict;
```
Comment (aab-odoo) [inline addons/web/static/src/js/services/data_manager.js]: not sure that options should be in fields, it is probably only necessary in fieldsInfo

File: addons/web/static/src/js/services/data_manager.js [hunk 2]
```diff
@@ -406,8 +406,12 @@ return core.Class.extend({
                 if (fieldsInfo[node.attrs.name].fieldDependencies) {
                     var deps = fieldsInfo[node.attrs.name].fieldDependencies;
                     for (var dependency_name in deps) {
+                        var dependency_dict = {'name': dependency_name, 'type': deps[dependency_name].type, 'options':  deps[dependency_name].options || {}};
```
Comment (aab-odoo) [inline addons/web/static/src/js/services/data_manager.js]: no need to quote keys in dict definitions in JS

**Comment by kebeclibre** [comment]
@aab-odoo 
Test finished and efficient: it does crash when my fix is absent, both in saas-16 and v11

File: addons/web/static/tests/fields/basic_fields_tests.js [hunk 3]
```diff
@@ -1464,6 +1464,56 @@ QUnit.module('basic_fields', {
...
+                        '</tree>' +
+                        '<form>' +
+                            '<field name="image" widget="image"/>' +
+                        '</form>' +
+                    '</field>' +
+                '</form>',
+            res_id: 1,
+            mockRPC: function (route, args) {
+                if (route === 'data:image/png;base64,myimage' || route === 'data:image/png;base64,product_image') {
+                    assert.ok(true, "should called the correct route");
```
Comment (aab-odoo) [inline addons/web/static/tests/fields/basic_fields_tests.js]: just fyi, assert.step() and assert.verifySteps() are useful for that kind of situations

File: addons/web/static/tests/fields/basic_fields_tests.js [hunk 4]
```diff
@@ -1464,6 +1464,56 @@ QUnit.module('basic_fields', {
...
+                    return $.when('wow');
+                }
+                return this._super.apply(this, arguments);
+            },
+        });
+
+        assert.strictEqual(form.$('tr.o_data_row').length, 1, 'There should be one record in the many2many');
```
Comment (aab-odoo) [inline addons/web/static/tests/fields/basic_fields_tests.js]: could you use the same style as other tests, i.e. limit to 80 (soft limit) or 100 (hard limit) characters? Typically, we write the last argument at the next line.

File: addons/web/static/tests/fields/basic_fields_tests.js [hunk 5]
```diff
@@ -1464,6 +1464,56 @@ QUnit.module('basic_fields', {
...
+
+        var timmy_field = form.renderer.allFieldWidgets.partner_1[1].record.fieldsInfo.form.timmy;
```
Comment (aab-odoo) [inline addons/web/static/tests/fields/basic_fields_tests.js]: this is not ok. Tests should not do introspection (in most cases). You can test what is in the DOM, and what is sent to the server by RPCs. In general, it is enough.

File: addons/web/static/tests/fields/basic_fields_tests.js (hunk 3 above)
Comment (aab-odoo) [reply addons/web/static/tests/fields/basic_fields_tests.js]: "should call the correct route"

Comment (aab-odoo) [reply addons/web/static/tests/fields/basic_fields_tests.js]: according to the number of asserts, only one of those routes is called, so you should simplify the condition

File: addons/web/static/src/js/services/data_manager.js (hunk 1 above)
Comment (aab-odoo) [reply addons/web/static/src/js/services/data_manager.js]: The other fields (in the fields object, not fieldsInfo) don't have a 'options' key, so i believe that my above comment is relevant. I'd like to understand why it crashes without it.

File: addons/web/static/src/js/services/data_manager.js (hunk 2 above)
Comment (aab-odoo) [reply addons/web/static/src/js/services/data_manager.js]: you already edited the line, so you can remove them...

File: addons/web/static/tests/fields/basic_fields_tests.js (hunk 3 above)
Comment (kebeclibre) [reply addons/web/static/tests/fields/basic_fields_tests.js]: for the options that shouldn't be in fields, I misread your comment, sorry :)

**Comment by kebeclibre** [comment]
@aab-odoo 
I made the changes

File: addons/web/static/src/js/services/data_manager.js [hunk 6]
```diff
@@ -406,8 +406,12 @@ return core.Class.extend({
                 if (fieldsInfo[node.attrs.name].fieldDependencies) {
                     var deps = fieldsInfo[node.attrs.name].fieldDependencies;
                     for (var dependency_name in deps) {
+                        var dependency_dict = {name: dependency_name, type: deps[dependency_name].type};
                         if (!(dependency_name in fieldsInfo)) {
-                            fieldsInfo[dependency_name] = {'name': dependency_name, 'type': deps[dependency_name].type, 'options':  deps[dependency_name].options || {}};
+                            fieldsInfo[dependency_name] = _.extend(dependency_dict, {options: deps[dependency_name].options || {}});
```
Comment (aab-odoo) [inline addons/web/static/src/js/services/data_manager.js]: _.extend({}, dependency_dict, ...) otherwise, you're updating dependency_dict in place

**Pull request closed without merging**